tests/lambda/lambda_small_step.rs

```diff
@@ -3,51 +3,68 @@ use crate::lambda::*;
+        let mut future_steps = Vec::new();
+        for cand in crate::lambda::big_step::candidates(eg) {
+            let app_id = eg.lookup(&cand.app).unwrap();
 
-        // L0 = Lambda::App(l, t).slots() -- "the root level"
-        // t.slots(), l.slots(), app_id.slots() :: L0
+            // L0 = Lambda::App(l, t).slots() -- "the root level"
+            // t.slots(), l.slots(), app_id.slots() :: L0
 
-        // L1 = Lambda::Lam(x, b).slots() = slots(l.id)
+            // L1 = Lambda::Lam(x, b).slots() = slots(l.id)
 
-        let Lambda::App(l, t) = cand.app.clone() else {
-            panic!()
-        };
-        let Lambda::Lam(Bind { slot: x, elem: b }) = cand.lam.clone() else {
-            panic!()
-        };
-        assert_eq!(x, Slot::numeric(0));
+            let Lambda::App(l, t) = cand.app.clone() else {
+                panic!()
+            };
+            let Lambda::Lam(Bind { slot: x, elem: b }) = cand.lam.clone() else {
+                panic!()
+            };
+            assert_eq!(x, Slot::numeric(0));
 
-        // b.m :: slots(b.id) -> L1
-        // l.m :: slots(l.id) -> L0 (and thus L1 -> L0)
+            // b.m :: slots(b.id) -> L1
+            // l.m :: slots(l.id) -> L0 (and thus L1 -> L0)
 
-        // The L0-equivalent of x.
-        let x_root = Slot::fresh();
+            // The L0-equivalent of x.
+            let x_root = Slot::fresh();
 
-        let mut l_m = l.m.clone();
-        l_m.insert(x, x_root);
-        let b = b.apply_slotmap(&l_m);
+            let mut l_m = l.m.clone();
+            l_m.insert(x, x_root);
+            let b = b.apply_slotmap(&l_m);
 
-        assert!(eg.is_alive(b.id));
-        for b_node in eg.enodes_applied(&b) {
-            let new = step(x_root, t.clone(), &b_node, eg);
-            future_unions.push((new, app_id.clone()));
+            assert!(eg.is_alive(b.id));
+            for b_node in eg.enodes_applied(&b) {
+                future_steps.push((app_id.clone(), x_root, t.clone(), b_node.clone()));
+                // let new = step(x_root, t.clone(), &b_node, eg);
+                // future_unions.push((new, app_id.clone()));
+            }
         }
+        future_steps
+    });
+    let applier = Box::new(
+        |future_steps: Vec<(AppliedId, Slot, AppliedId, Lambda)>, eg: &mut EGraph<Lambda>| {
+            let mut future_unions = Vec::new();
+            for (app_id, x_root, t, b_node) in future_steps {
+                let new = step(x_root, t.clone(), &b_node, eg);
+                future_unions.push((new, app_id.clone()));
+            }
+            for (x, y) in future_unions {
+                eg.union_justified(&x, &y, Some("beta-rewrite-small-step".to_string()));
+            }
+        },
+    );
+    vec![RewriteT {
+        searcher: searcher,
+        applier: applier,
     }
-
-    for (x, y) in future_unions {
-        eg.union_justified(&x, &y, Some("beta-rewrite-small-step".to_string()));
-    }
+    .into()]
```

Comment:
here too, I would have intuitively expected the "step" function to be part of the searcher.